### PR TITLE
Feature/add forms

### PR DESCRIPTION
**Pull request by janani-ravikumar**
added template-driven form for user login
added reactive form for edit user info

### CODE REVIEW

File: apm/src/app/user/login/login.component.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  styleUrls: ['./login.component.css']
+})
+
+export class LoginComponent{
+  userName: any;
+  password: any; 
+  homepageUrl: string = "/welcome";
+  constructor(private userService: UserService, private router: Router) { }
+
+  login(formValues: any) {
```
Comment (mukesh-jangir) [inline apm/src/app/user/login/login.component.ts:16]: Since we use typescript to help us avoid errors related to dynamic typing in javascript, we should avoid using any as a type. That does not help. Instead you can create an interface for the data you are expecting and use that

File: apm/src/app/user/login/login.component.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+@Component({
+  templateUrl: './login.component.html',
+  styleUrls: ['./login.component.css']
+})
+
+export class LoginComponent{
+  userName: any;
+  password: any; 
+  homepageUrl: string = "/welcome";
```
Comment (mukesh-jangir) [inline apm/src/app/user/login/login.component.ts:13]: just a suggestion, typescript is smart enough to conclude the type when a value is assigned, so we can ignore using the type here.

File: apm/src/app/user/user.service.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { Injectable } from '@angular/core';
+import { IUser } from './user';
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class UserService {
+
+  currentUser: IUser | undefined;
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:10]: I would use `IUser | null` but that is my preference. Undefined in javascript is something that is not even defined, but we have defined the variable and so it should have some value. Again this is just a suggestion and something that I advocate.

File: apm/src/app/user/user.service.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  login(userName: string, password: string) {
+    this.currentUser = {
+      id : 1,
+      firstName: userName,
+      lastName: ''
+    }
+  }
+
+  isAuthenticated() {
+    return !!this.currentUser;
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:21]: I have used this myself, but over the years the suggestion that I have received is to use `Boolean(this.currentUser)`

File: apm/src/app/user/login/login.component.ts (hunk 1 above)
Comment (janani-ravikumar) [reply apm/src/app/user/login/login.component.ts:16]: Okay, I will create an interface.

File: apm/src/app/user/user.component.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  public editForm!: FormGroup;
+  private firstName!: FormControl;
+  private lastName!: FormControl;
+  private homepageUrl = "/welcome";
+  constructor(private userService: UserService, private router: Router) { }
+  
+  ngOnInit(): void {
+    this.firstName = new FormControl(
+      this.userService.currentUser?.firstName, Validators.required
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:21]: `this.userService.currentUser?.firstName` can have a value of undefined potentially. I mean the code is also indicating towards that. you can use `this.userService.currentUser?.firstName ?? ''` instead and make sure you are assigning an empty string when the left expression evaluates into null or undefined

Alternatively you can also use `this.userService.currentUser?.firstName || ''` but that is not accurate

example
0 || 'abc'  gives 'abc' because 0 is falsy value
0 ?? 'abc' gives 0 because it is not null or undefined

File: apm/src/app/user/user.component.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  validateFirstName() {
+    return this.firstName.valid || this.firstName.untouched;
+  }
+
+  validateLastName() {
+    return this.lastName.valid;
+  }
+
+  UpdateUserInfo(formValues: IUser) {
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:45]: you do not need to get the formValues from the template

File: apm/src/app/user/user.service.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+
+export class UserService {
+
+  public currentUser!: IUser | null;
+    
+  login(userName: string, password: string) {
+    this.currentUser = {
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:13]: .

File: apm/src/app/user/user.service.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      lastName: ''
+    }
+  }
+
+  isAuthenticated() {
+    return Boolean(this.currentUser);
+  }
+
+  UpdateInfo(firstName: string, lastName: string) {
+    this.currentUser = {
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:25]: do we need to create a new instance here ?

File: apm/src/app/user/user.service.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { Injectable } from '@angular/core';
+import { IUser } from './user';
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class UserService {
+
+  public currentUser!: IUser | null;
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:10]: can you default this to null, otherwise undefined is a probable value still.

File: apm/src/app/welcome/navigation.component.html [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<nav class='navbar navbar-light bg-light'>
+    <a class='navbar-brand'>{{pageTitle}}</a>
+    <ul class='nav nav-pills'>
+      <li><a class='nav-link' routerLink='/welcome'>Home</a></li>
+      <li><a class='nav-link' routerLink='/products'>Product List</a></li>
+      <li *ngIf="!userService.isAuthenticated()"><a class='nav-link' routerLink='/login'>Login</a></li>
```
Comment (mukesh-jangir) [inline apm/src/app/welcome/navigation.component.html:6]: same expression is evaluated 3 times. Let me know if you have read about ng-template and ng-container

**Review by mukesh-jangir — COMMENTED** [review]
Overall it is good to see a PR and comment on it. Kudos!

File: apm/src/app/user/login/login.component.html [hunk 11]
```diff
@@ -0,0 +1,18 @@
+<h1>Login</h1>
+<hr>
+<div class="col-md-4">
+  <form autocomplete="off" #loginForm='ngForm' (ngSubmit)='login(loginForm.value)' novalidate>
+    <div class="form-group" >
+      <label for="userName">User Name:</label>
+      <input (ngModel)='userName' name='userName' id="userName" type="text" class="form-control" required placeholder="User Name..." />
```
[the remark this answers is not in the document]
Comment (janani-ravikumar) [reply apm/src/app/user/login/login.component.html:7]: The (ngModel) attribute has to be changed with 'loginName' so that it gets bind to form value

File: apm/src/app/welcome/navigation.component.html (hunk 10 above)
Comment (janani-ravikumar) [reply apm/src/app/welcome/navigation.component.html:6]: Not yet. I will update this after I read about ng-template and ng-container

File: apm/src/app/user/login/login.component.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (mukesh-jangir) [reply apm/src/app/user/login/login.component.html:7]: you do not have to change the ngModel attribute. try changing the name alone and see what loginForm.value would give you

[the remark this answers is not in the document]
Comment (janani-ravikumar) [reply apm/src/app/user/login/login.component.html:7]: Yes, I tried. It gives the object with field Name as loginName and not the username and password fields.

File: apm/src/app/user/user.service.ts (hunk 7 above)
Comment (janani-ravikumar) [reply apm/src/app/user/user.service.ts:13]: I didn't get what you meant here. Could you please brief on this?

Comment (mukesh-jangir) [reply apm/src/app/user/user.service.ts:13]: nothing, I did not know how to delete the comment, so simply put it as `.`

File: apm/src/app/user/user.component.html [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<h1>Edit Profile</h1>
+<hr>
+<div class="col-md-4">
+  <form autocomplete="off" [formGroup]="editForm" (ngSubmit)="UpdateUserInfo()" novalidate>
+    <div class="form-group">
+      <label for="firstName">First Name:</label>
+      <input formControlName ="firstName" name="firstName" id="firstName" type="text" class="form-control" placeholder="First Name..." />
+      <em *ngIf="!FirstName">Required</em>
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.html:8]: I believe this is for showing the validation, so the naming could be something like isFirstNameValid or isFirstNameBlank

File: apm/src/app/user/user.component.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+})
+export class UserComponent implements OnInit {
+
+  public editForm!: FormGroup;
+  private firstName!: FormControl;
+  private lastName!: FormControl;
+  private readonly homepageUrl = "/welcome";
+  constructor(private readonly userService: UserService, private readonly router: Router) { }
+  
+  public get FirstName() {
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:18]: always specify the return types for getters and methods, that way you have typescript to indicate to you if you are returning something that is not expected from them

File: apm/src/app/user/user.component.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    return this.firstName.value;
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:19]: Also you should always check for what you want to test against, like are you testing for a blank value or null or undefined. With the way it is we are still depending or thinking of implementation like it is dynamically typed

File: apm/src/app/user/user.component.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+import { Router } from '@angular/router';
+import { UserService } from './user.service';
+
+@Component({
+  templateUrl: './user.component.html',
+  styleUrls: ['./user.component.css']
+})
+export class UserComponent implements OnInit {
+
+  public editForm!: FormGroup;
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:12]: try to avoid using `!` anywhere. The value for this technically speaking (without considering angular internal working) is undefined unless the ngOnInit method is called explicitly. So although we are saying the the editForm is of type `FormGroup` it actually is `FormGroup | undefined`

Comment (mukesh-jangir) [reply apm/src/app/user/user.component.ts:12]: Having all that said, please do not take all that what i say as something to simply do, Someplaces it is also a development convention that people have considering the internal working of Angular

File: apm/src/app/user/user.component.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  public cancel() {
+    this.router.navigate([this.homepageUrl]);
+  }
+
+  public UpdateUserInfo() {
+    if (this.editForm.valid) {
+      var editFormValue = this.editForm.value;
+      this.userService.UpdateInfo(editFormValue?.firstName, editFormValue?.lastName)
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:47]: why is `?` used for `editFormValue?.firstName` ?

File: apm/src/app/user/user.component.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+      this.router.navigateByUrl(this.homepageUrl)
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.component.ts:48]: I see that you have used navigate at one place and navigateByUrl here, was that intentional ?

File: apm/src/app/user/user.module.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    CommonModule,
+    FormsModule,
+    ReactiveFormsModule,
+    RouterModule.forChild([
+      { path: 'login', component: LoginComponent},
+      { path: 'edit', component: UserComponent}
+    ])
+  ],
+  providers: [
+    UserService
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.module.ts:24]: Just a note here again, for this service it does not matter if you had repeated the service registeration here and on the appModule. That is because the module is not lazy loaded and Angular will provide the same instance. But it would be different in case a module was lazy loaded. Just something to have in mind when you learn about lazy loading modules

File: apm/src/app/user/user.service.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+  public login(userName: string, password: string) {
+    this.currentUser = {
+      id : 1,
+      firstName: userName,
+      lastName: ''
+    }
+  }
+
+  public isAuthenticated() {
+    return Boolean(this.currentUser);
```
Comment (mukesh-jangir) [inline apm/src/app/user/user.service.ts:19]: since we have modified the type of currentUser, I think we need to check for it to be null ?

Comment (mukesh-jangir) [reply apm/src/app/user/user.service.ts:19]: like `return this.currentUser !== null`

**Review by mukesh-jangir — APPROVED** [review]
Great work on Angular forms and reactive forms

File: apm/src/app/user/user.component.ts (hunk 17 above)
Comment (janani-ravikumar) [reply apm/src/app/user/user.component.ts:48]: Yes, was just trying both the commands

File: apm/src/app/user/user.component.ts (hunk 16 above)
Comment (janani-ravikumar) [reply apm/src/app/user/user.component.ts:47]: I just checked, it didn't make any difference, will remove that.
